bson/datetime_ms.py

```diff
@@ -127,24 +129,33 @@ def _millis_to_datetime(millis: int, opts: CodecOptions) -> Union[datetime.datet
         or opts.datetime_conversion == DatetimeConversion.DATETIME_CLAMP
         or opts.datetime_conversion == DatetimeConversion.DATETIME_AUTO
     ):
-        tz = opts.tzinfo or datetime.timezone.utc
-        if opts.datetime_conversion == DatetimeConversion.DATETIME_CLAMP:
-            millis = max(_min_datetime_ms(tz), min(millis, _max_datetime_ms(tz)))
-        elif opts.datetime_conversion == DatetimeConversion.DATETIME_AUTO:
-            if not (_min_datetime_ms(tz) <= millis <= _max_datetime_ms(tz)):
-                return DatetimeMS(millis)
-
-        diff = ((millis % 1000) + 1000) % 1000
-        seconds = (millis - diff) // 1000
-        micros = diff * 1000
-
-        if opts.tz_aware:
-            dt = EPOCH_AWARE + datetime.timedelta(seconds=seconds, microseconds=micros)
-            if opts.tzinfo:
-                dt = dt.astimezone(tz)
-            return dt
-        else:
-            return EPOCH_NAIVE + datetime.timedelta(seconds=seconds, microseconds=micros)
+        try:
+            tz = opts.tzinfo or datetime.timezone.utc
+            if opts.datetime_conversion == DatetimeConversion.DATETIME_CLAMP:
+                millis = max(_min_datetime_ms(tz), min(millis, _max_datetime_ms(tz)))
+            elif opts.datetime_conversion == DatetimeConversion.DATETIME_AUTO:
+                if not (_min_datetime_ms(tz) <= millis <= _max_datetime_ms(tz)):
+                    return DatetimeMS(millis)
+
+            diff = ((millis % 1000) + 1000) % 1000
+            seconds = (millis - diff) // 1000
+            micros = diff * 1000
+
+            if opts.tz_aware:
+                dt = EPOCH_AWARE + datetime.timedelta(seconds=seconds, microseconds=micros)
+                if opts.tzinfo:
+                    dt = dt.astimezone(tz)
+                return dt
+            else:
+                return EPOCH_NAIVE + datetime.timedelta(seconds=seconds, microseconds=micros)
+        except Exception:
```

Comment:
We should only catch `InvalidBSON`

Reply:
Ah, so the thing here is this exception is actually an `OverflowError` that gets thrown at this point in the code, so catching only `InvalidBSON` will not trigger the exception capture. Further down the pipe it gets re-converted into `InvalidBSON`. We run the same/similar behavior in the calling function [`_bson_to_dict`](https://github.com/mongodb/mongo-python-driver/blob/master/bson/__init__.py#L594-L607). 
```
def _bson_to_dict(data: Any, opts: CodecOptions) -> Any:
    """Decode a BSON string to document_class."""
    data, view = get_data_and_view(data)
    try:
        if _raw_document_class(opts.document_class):
            return opts.document_class(data, opts)
        _, end = _get_object_size(data, 0, len(data))
        return _elements_to_dict(data, view, 4, end, opts)
    except InvalidBSON:
        raise
    except Exception:
        # Change exception type to InvalidBSON but preserve traceback.
        _, exc_value, exc_tb = sys.exc_info()
        raise InvalidBSON(str(exc_value)).with_traceback(exc_tb)
```

Based on the criteria we set, though, I can definitely remove using `Exception` and I think favor using something still fairly broad, but clearly more specific. 

e.g. `OverflowError`'s base class `ArithmeticError`.